### ISSUE TITLE
Siddhesh GSoC 2024 status updates

### ISSUE THREAD

**Issue by siddhesh195** [animint/animint2#133]
In this thread I will post status of various tasks I am doing as a part of GSoC 2024.

**Comment by siddhesh195** [comment]
Date: 06/18/2024

- chromote testing task : Everything is done and the PR(#126) is ready to be reviewed. All the failing tests have been migrated to chromote and GitHub actions workflow has been switched to use chromote.
- capture screenshot in animint2pages: initialized chromote session in animint2pages function. chromote will be used to capture the screenshot. Trying to solve the test error (https://github.com/animint/animint2/issues/129#issuecomment-2179545850)

**Comment by siddhesh195** [comment]
Date: 06/19/2024

- chromote testing task : Everything is done and the PR(#126) is ready to be reviewed. All the failing tests have been migrated to chromote and GitHub actions workflow has been switched to use chromote.
- capture screenshot in animint2pages:
   1.   initialized chromote session in animint2pages function. 

    2.   chromote will be used to capture the screenshot. Trying to solve the test error (https://github.com/animint/animint2/issues/129#issuecomment-2179545850)
     3. trying to solve the issue of changes in animint2pages function in z_pages.R  not affecting the test compiler-ghpages which use it. 
     4. had implemented the code of saving the screenshot locally few days ago but somehow lost track of it. after solving issue mentioned in point 3 will be able to do that that again.

**Comment by siddhesh195** [comment]
Date: 06/20/2024

- chromote testing task : Everything is done and the PR(https://github.com/animint/animint2/pull/126) is ready to be reviewed. All the failing tests have been migrated to chromote and GitHub actions workflow has been switched to use chromote.

- capture screenshot in animint2pages:

   1)initialized chromote session in animint2pages function.

   2)chromote will be used to capture the screenshot. Trying to solve the test error (https://github.com/animint/animint2/issues/129#issuecomment-2179545850)

   3)trying to solve the issue of changes in animint2pages function in z_pages.R not affecting the test compiler-ghpages which use it.
   
 4) added code to take  screenshot using Chromote and save that as png file in temporary local directory which also has animint

**Comment by siddhesh195** [comment]
Date: 06/20/2024

chromote testing task : Everything is done and the PR(https://github.com/animint/animint2/pull/126) is ready to be reviewed. All the failing tests have been migrated to chromote and GitHub actions workflow has been switched to use chromote.

capture screenshot in animint2pages:

  1)initialized chromote session in animint2pages function.

  2)chromote will be used to capture the screenshot. Trying to solve the test error (https://github.com/animint/animint2/issues/129#issuecomment-2179545850)

  3) added code to take screenshot using Chromote and save that as png file in temporary local directory which also has animint. working on verifying whether the screenshot is saved locally.

**Comment by siddhesh195** [comment]
Date: 06/22/2024

chromote testing task : Everything is done and the PR(https://github.com/animint/animint2/pull/126) is ready to be reviewed. All the failing tests have been migrated to chromote and GitHub actions workflow has been switched to use chromote.

capture screenshot in animint2pages:

  1)initialized chromote session in animint2pages function.

  2)chromote will be used to capture the screenshot. Trying to solve the test error (https://github.com/animint/animint2/issues/129#issuecomment-2179545850)

  3) added code to take screenshot using Chromote and save that as png file in temporary local directory which also has animint. working on verifying whether the screenshot is saved locally.

  4) comparing chrome sessions screenshot API and chromote's screenshot API to see if one is better option over another

**Comment by siddhesh195** [comment]
Date: 06/23/2024

chromote testing task : Everything is done and the PR(https://github.com/animint/animint2/pull/126) is ready to be reviewed. All the failing tests have been migrated to chromote and GitHub actions workflow has been switched to use chromote.

capture screenshot in animint2pages:





  1)added code to take screenshot using Chromote and save that as png file in temporary local directory which also has animint. working on verifying whether the screenshot is saved locally.

  2)comparing chrome sessions screenshot API and chromote's screenshot API to see if one is better option over another
 3) wrote a test to verify screenshot exists on GitHub pages and ensured that the test passes locally on my machine.

**Comment by tdhock** [comment]
hi thanks for the updates, but please avoid writing the same thing as the previous day (instead please just write what is new)

**Comment by siddhesh195** [comment]
Date: 06/24/2024

Solved failures on GitHub actions of commits to send screenshot to GitHub pages.

**Comment by siddhesh195** [comment]
Date: 06/25/2024

For chromote testing PR (#126 ), started the work of removing codes in testthat directory for browsers other than chromote as a part of final steps to fully migrate to chromote for testing.

**Comment by siddhesh195** [comment]
Date: 06/26/2024

For the screenshot task(PR #131 ), after verifying by a test that screenshot png file is there on GitHub pages, started the work to figure why screenshot  file is just white. reading more about dynamic html because that might be the issue why screenshot of ghpages test is not taken properly just using URL of temporary directory in chromote screenshot API

**Comment by siddhesh195** [comment]
Date: 06/27/2024

for PR #126 ,completed the work of deleting the code for browsers other than chromote.

**Comment by siddhesh195** [comment]
Date: 06/28/2024

PR #131 , solved the issue of blank screenshot by start a server which points to local directory which has dynamic html and taking screenshot by navigating to local host.

**Comment by siddhesh195** [comment]
Date: 06/29/2024

for PR #131  , the server pointing to temporary local directory to handle dynamic html for taking screenshot is working and screenshot is taken properly. but now I am facing error when I try to run the test again because when server tries to start again the port is busy because previous server has not been terminated. trying to figure out how to do that. I faced this error on my Mac and not on my windows. Maybe that's because windows is automatically terminating the inactive server process to free the port while Mac is not.

**Comment by siddhesh195** [comment]
Date: 06/30/2024

for PR #131  , had  a meeting with @nhintruong  and guided her to integrate magick package API calls for trimming whitespace in screenshot.

**Comment by siddhesh195** [comment]
> Date: 06/29/2024
> 
> for PR #131 , the server pointing to temporary local directory to handle dynamic html for taking screenshot is working and screenshot is taken properly. but now I am facing error when I try to run the test again because when server tries to start again the port is busy because previous server has not been terminated. trying to figure out how to do that. I faced this error on my Mac and not on my windows. Maybe that's because windows is automatically terminating the inactive server process to free the port while Mac is not.

Date: 07/1/2024

tried to solve port busy error by stopping non blocking R session using it's PID in text file generated by run_server function.

**Comment by tdhock** [comment]
just kill the server after the screenshot was taken?

**Comment by siddhesh195** [comment]
> just kill the server after the screenshot was taken?

@tdhock for some reason if I just use servr::httd(port = 8080, dir = normalized_directory), chromote is not able to navigate and gives timeout error. But if I pass this code in a new R session initialized by system2() then chromote is able to navigate to localhost url and then take the screenshot. So now I can only stop the server if I am able to terminate that entire new R session initialized for the server. I am trying to do that. But since tcp port management is done by OS, overall it would be complex to implement a portable code for that I think. I get tcp port busy error only on Mac, not on windows. So for now I am just assigning random available ports compatible with chrome using a function in servr package which does that.

**Comment by tdhock** [comment]
you should file an issue with servr package
a work-around is to maybe keep one servr for the entire R session, instead of creating a new one every time you run animint2pages
or using a random/different port instead of the same one every time

**Comment by siddhesh195** [comment]
> you should file an issue with servr package a work-around is to maybe keep one servr for the entire R session, instead of creating a new one every time you run animint2pages or using a random/different port instead of the same one every time

 The approach of using random available port worked. I will also try the new approach of keeping a same server.  thanks

**Comment by siddhesh195** [comment]
> you should file an issue with servr package a work-around is to maybe keep one servr for the entire R session, instead of creating a new one every time you run animint2pages or using a random/different port instead of the same one every time

@tdhock I tried two approaches for keeping a same server for the entire session. In the first approach I simply copied contents of temporary directory generated by animint2dir into the directory of server. In the second approach I set the already exisitng directory argument of animint2dir from Null to server directory and made changes inside animint2dir for it to work. In both approaches all test cases passed on my machine. The second passed on Github too. But the screenshot on GitHub pages is actually that of sever error message which says directory cannot be changed.

**Comment by tdhock** [comment]
if you want to test the content of the screenshot, you should test the content of the web page right before/after you take the screenshot (in addition to testing presence of file)

**Comment by siddhesh195** [comment]
Date: 07/2/2024

Instead of trying stopping the server's R session using PID in text file generated by server function started a new approach which does not try to kill existing server sessions but assign random available ports

**Comment by siddhesh195** [comment]
Date: 07/3/2024

Finished and tested the random server approach for correctness. Started working on code which fixed output directory of animint2dir

**Comment by siddhesh195** [comment]
Date: 07/4/2024

Finished the two approaches for persistent server, one is copying files from a temporary random directory of animint2dir to fixed directory of server. Another is passing an argument in animint2dir to fixed a directory.

**Comment by siddhesh195** [comment]
Date: 07/5/2024

Screenshot file was getting uploaded to GitHub pages using persistent server approaches but it was "not able to change directory" error message instead of screenshot of the plot. Started working to solve that.

**Comment by siddhesh195** [comment]
Date: 07/5/2024

PR #131 ,Fixed error while using persistent server, updated news and changed version

**Comment by siddhesh195** [comment]
Date: 07/7/2024

for html plot task starting ramping up on it by revising my javascript knowledge

**Comment by siddhesh195** [comment]
Date: 07/8/2024

for html plot task (issue #115 ), wrote a new test to add 3 plots to see how they are rendered in current implementation

**Comment by siddhesh195** [comment]
> Date: 07/8/2024
> 
> for html plot task (issue #115 ), wrote a new test to add 3 plots to see how they are rendered in current implementation

Date: 07/9/2024
for html plot task (issue #115 ), analyzed the html code of test written to ramp myself up for the task. Also made new commits for PR #126 related to refactoring.

**Comment by siddhesh195** [comment]
Date: 07/10/2024

opened new PR #139 for html plot task(issue #115 ). started looking at and trying to understand the code of animint.js to ramp myself up for the task

**Comment by siddhesh195** [comment]
Date: 07/11/2024

For discussing ggplot html table task (PR #115 ), Had a meeting with @Faye-yufan. Made commits to refactor code for PR #126

**Comment by siddhesh195** [comment]
Date: 07/12/2024

For PR #126 , refactored the code and started the work to add a function for Runtime$evaluate API call.

**Comment by siddhesh195** [comment]
> Date: 07/12/2024
> 
> For PR #126 , refactored the code and started the work to add a function for Runtime$evaluate API call.

Date: 07/13/2024

finished the work to add a helper function for Runtime$evaluate

**Comment by siddhesh195** [comment]
> > Date: 07/12/2024
> > For PR #126 , refactored the code and started the work to add a function for Runtime$evaluate API call.
> 
> Date: 07/13/2024
> 
> finished the work to add a helper function for Runtime$evaluate

Date: 07/14/2024

refactored the code and ensured the all tests pass locally. Started the work to try solve the issue on GitHub Actions that checks are running for hours and never stop. Or sometimes they stop and give errors.

**Comment by siddhesh195** [comment]
> > > Date: 07/12/2024
> > > For PR #126 , refactored the code and started the work to add a function for Runtime$evaluate API call.
> > 
> > 
> > Date: 07/13/2024
> > finished the work to add a helper function for Runtime$evaluate
> 
> Date: 07/14/2024
> 
> refactored the code and ensured the all tests pass locally. Started the work to try solve the issue on GitHub Actions that checks are running for hours and never stop. Or sometimes they stop and give errors.

Date: 07/15/2024

for chromote testing (PR #126), figured out why checks were failing on GitHub and rectified that issue by reverting commit which removed packages in suggests in .yaml file. for ggplot table task (#139 ), resumed the work of understanding how functions in animint.js file work. to ramp up, started making temporary changes in code and see how they change the plots to understand how different functions work.

**Comment by siddhesh195** [comment]
Date: 07/16/2024

for ggplot table task (#139 ), added the code which puts all plots in a single row of newly created outer table. this fixes relative positions of plots despite interaction with the browser. will build upon this logic in next tasks to fully implement the table functionality. only one test "knit-print" failing on GitHub actions but that was failing for current phantomjs setup for chromote testing task too (#126)

**Comment by tdhock** [comment]
thanks for the update.
also please update the chromote testing pr, I would like to merge that soon.

**Comment by siddhesh195** [comment]
> thanks for the update. also please update the chromote testing pr, I would like to merge that soon.

@tdhock yes. I have updated that PR. please let me know if I have missed something.

**Comment by siddhesh195** [comment]
Date: 07/22/2024

These are the new updates for the last few days.

for chromote testing(#126 ),
1) done the refactoring
2) solved the issue of GitHub actions running perpetually by adding sleep in angle axis rotate test.
3) waiting for further feedback from @tdhock and @Faye-yufan 


for screenshot task (#131  ), explained to @Faye-yufan my approach for persistent server for handling dynamic html 

for html table plot task(#139 ), I have added code to add maximum 2 plots per row of newly created outer table and then make a new row. this code is just a middle step to help me understand how this can potentially work. to complete the task,I will build upon this implementation if it is the most suitable approach.

**Comment by siddhesh195** [comment]
> Date: 07/22/2024
> 
> These are the new updates for the last few days.
> 
> for chromote testing(#126 ),
> 
> 1. done the refactoring
> 2. solved the issue of GitHub actions running perpetually by adding sleep in angle axis rotate test.
> 3. waiting for further feedback from @tdhock and @Faye-yufan
> 
> for screenshot task (#131 ), explained to @Faye-yufan my approach for persistent server for handling dynamic html
> 
> for html table plot task(#139 ), I have added code to add maximum 2 plots per row of newly created outer table and then make a new row. the relative positions of all plots are now fixed regardless of how the user interacts with the browser. this code is just a middle step to help me understand how this can potentially work. to complete the task,I will build upon this implementation if it is the most suitable approach to move forward

**Comment by siddhesh195** [comment]
Date: 07/24/2024

for html table plot task(#139 ), in the previous implementation of fixed table, I had only fixed an upper limit for number of plots for each row and it used to create a new row in the newly created outer table when the current row was filled.

Now I added code to put plots in specific location of the outer table according to user given values of row and col. this code is in rudimentary form and for now it is only working for plots file I have made. but I got an understanding of a possible way I can go forward to complete this task.

I have also analyzed the theme_animint function and the third party theme function to ramp up. because as @tdhock had said in issue #115, the arguments for row and col should ultimately go as parameters to theme_animint function, if I am interpreting that correctly. 

the screenshot for the table implementation is below. I could not capture the entire screenshot despite zooming out. it's a 3 by 3 table with 4 plots in it. the location of the plots was selected by me and to put them in right place I used html IDs for each unique cell comprising of it's row and column number.
<img width="927" alt="Screenshot 2024-07-24 at 6 05 24 PM" src="https://github.com/user-attachments/assets/b4d477ad-0dc5-434b-acba-4ce2b21281a4">

**Comment by siddhesh195** [comment]
Date: 07/25/2024

these are the updates for html plot task(#139 ), 

I have not made any commits of the recent progress so far because the code is in rudimentary form and for now it is only compatible if we pass row and column parameters in theme.

here are the new changes I implemented today,

1) As @Faye-yufan suggested I made the color of the outer table border as solid white, for some reason if I do not keep the borders, the layout of plots is not proper

2) I modified the theme function so that it accepts row and column as valid input parameters. I also had to modify z_animint file so that it puts those row and column parameters in json file.

3) I added code in animint.js to see maximum row and column number and construct outer table with that dimensions and each plot is put in the right place of that outer table according to row and column number that the user has passed using theme function.


These are the changes I need to do soon before trying to implement more complex functionalities like row span:

1) I need to make changes in the source code soon to bring back the option of plotting if the user does not want the option of fixed user defined location for the plots. for that I can use the previous implementation of fixing some specific number of plots per row so that the browser interactivity would still not affect the relative locations of the plots.

2) In a meeting with @Faye-yufan today, I asked her one doubt that I had. The doubt was that if the user is not consistent and puts row and column numbers for some plots and not for the others, then what do we do ? I am wondering if we can first put the  plots who want a dedicated location in the outer table and then put rest of them in all other empty cells. I will try doing that. but yufan suggested to first implement an easier solution of simply giving an error message to the user if the user is inconsistent regarding passing row and column parameter using theme function.


Below are the screenshots of the latest implementations I have mentioned in this status update  : 

screenshot of passing row and column into theme function :

<img width="553" alt="Screenshot 2024-07-25 at 9 39 27 PM" src="https://github.com/user-attachments/assets/40caa703-c1c5-4170-b945-164740aacd29">

screenshot of the table with plots in locations determined by theme function arguments :

<img width="786" alt="Screenshot 2024-07-25 at 9 16 33 PM" src="https://github.com/user-attachments/assets/4949c3bf-30b1-413d-baee-1d4b17ea5d03">

**Comment by siddhesh195** [comment]
Date: 07/26/2024

these are the updates for html plot task(https://github.com/animint/animint2/pull/139 ),

1) implemented the the code to run tests if user if not passing any row or column argument in theme. a outer table will be created with 3 maximum plots per rows and then a new row. this is done to fixed relative positions of plots regardless interactivity even if the user does not want the option of putting plot in specific location of outer table.

2) wrote a test to verify that that the changes mentioned above work

3) but for now if the user wants the option of specific location for the plot, the user has to do that for each and every plot. I am yet to add code to show error if the user is inconsistent in giving row and column arguments.

my immediate next goals for this task,

1) add error message mentioned in point 3

2) make theme_animint function work like theme function is working right now

3) try to implement functionality which handles proceeding with plotting even if user is inconsistent regarding adding rows and columns in theme or theme_animint

After today's work I was finally able to make a commit because the code will now work even if user does not pass row and column in any plots.

**Comment by siddhesh195** [comment]
@tdhock @Faye-yufan 

html table task (#139 ) is at this stage right now -

1) using theme_animint function to pass row and column parameters

2) if the user adds only row or col then give error in the R session

3) construct an outer html table with minimum size such that it have the largest row and column passed by the user. if all plots together require more size than this minimum size then the predetermined size of the table is increased accordingly

4) put plots with dedicated locations first and then put rest of the plots in available empty positions. to maintain linear time complexity while doing that, I maintain pointers of row and column that keep moving ahead until a new empty location is found


what I am yet to do:

1) add functionality of row span and column span

2) ensure the code is neatly organized

3) think and discuss with you what more can be done for this task in terms of adding new functionality or changing the existing functionality. 

    Here are some examples I can think of-

   a) will there be some upper limit determined by the count of plots for the row or column number user enters.

   b) should there be options to the user like show grid or hide grid, choose background color of each individual cell etc.

**Comment by tdhock** [comment]
thanks
for now no grid options / background color is fine (just use white background)
not sure I understand what you mean by pointers? typically in JS we use arrays and integer indices for this: array[0], array[2], etc
not sure I undersand "upper limit" of what? typically there will only be 2-3 rows and columns.

**Comment by siddhesh195** [comment]
> thanks for now no grid options / background color is fine (just use white background) not sure I understand what you mean by pointers? typically in JS we use arrays and integer indices for this: array[0], array[2], etc not sure I undersand "upper limit" of what? typically there will only be 2-3 rows and columns.

@tdhock 

I should have clarified more what I meant by pointers, by pointers I meant var pointer_row and var pointer_column while putting plots with no dedicated locations, they keep track of where we are in table. so we do not have to start searching for empty spots from row 0 column 0 for each new plot.  so time complexity remains O(n) instead of O(n^2) 

by upper limit I mean should user be restricted for putting row and column number in theme_animint ? for example if the user has only few plots, say only 4 plots. should the user be able to pass very large number for row or column,  say row=100 and col=100 and make 10,000 cell table when there are only 4 plots ? do we need to keep some reasonable upper limit and prompt the user to keep row and column within that limit if it is exceeded ?

**Comment by tdhock** [comment]
yes good idea to error for row or col > number of plots

**Comment by siddhesh195** [comment]
hello @tdhock and @Faye-yufan 

GSoC 2024 tasks are in the following state:

1) Chromote Testing Merged (#126 ), I will continue to check #144 for follow up tasks

2) For plot layout task (#139 ), row and column can be passed in theme_animint function, for those plots where row and columns are not passed, the code handles that by putting plots in empty locations. but I missed a scenario where user passes negative values, I would add code to error if that happens. also, I will add more tests.  and I will refactor the code to remove any unnecessary blank lines, unnecessary indentations or  commented out code

3) for the screenshot task(#131 ), I am waiting for review from Toby.

4) Also, I have submitted the final submission form for GSoC 2024 . I have added link (https://siddhesh195.github.io/gsoc2024/)  of webpage where I am documenting the work I have done for GSoC 2024. Although I have added preview of all work I have done I am still updating that page to add more details about my tasks so that any future visitor would know better about the work.
